src/rules/switchDefaultRule.ts

```diff
@@ -15,41 +15,216 @@
+            const firstCases: string[] = cases.slice(0, -1).map(x => inspect(x));
+            const lastCase = inspect(cases[cases.length - 1]);
+
+            casesString = `cases \`${firstCases.join(", ")}\` and \`${lastCase}\``;
+        }
+        return `Switch statement is not exhaustive. Add the missing ${casesString} or a \`default\` case.`;
+    }
 
     public apply(sourceFile: ts.SourceFile): Lint.RuleFailure[] {
-        return this.applyWithFunction(sourceFile, walk);
+        const options = parseOptions(this.ruleArguments);
+
+        if (options.allowExhaustive) {
+            // allow-exhaustive is enabled, but we don't have type information
+            showWarningOnce(
+                Lint.Utils.dedent`
+                    ${this.ruleName} needs type info to use "${OPTION_ALLOW_EXHAUSTIVE}".
+                    See https://palantir.github.io/tslint/usage/type-checking/ for documentation on
+                    how to enable this feature.
+                `,
+            );
+
+            return [];
+        }
+
+        return this.applyWithFunction(sourceFile, walk, options);
+    }
+
+    public applyWithProgram(sourceFile: ts.SourceFile, program: ts.Program): Lint.RuleFailure[] {
+        return this.applyWithFunction(
+            sourceFile,
+            walk,
+            {
+                allowExhaustive: this.ruleArguments.indexOf(OPTION_ALLOW_EXHAUSTIVE) !== -1,
+            },
+            program.getTypeChecker(),
+        );
     }
 }
 
-function walk(ctx: Lint.WalkContext<void>) {
+function parseOptions(ruleArguments: any[]): Options {
+    return {
+        allowExhaustive: ruleArguments.indexOf(OPTION_ALLOW_EXHAUSTIVE) !== -1,
+    };
+}
+
+function walk(ctx: Lint.WalkContext<Options>, checker?: ts.TypeChecker) {
     return ts.forEachChild(ctx.sourceFile, function cb(node: ts.Node): void {
-        if (
-            node.kind === ts.SyntaxKind.SwitchStatement &&
-            !(node as ts.SwitchStatement).caseBlock.clauses.some(isDefaultClause)
-        ) {
-            ctx.addFailureAtNode(node, Rule.FAILURE_STRING);
+        if (isSwitchStatement(node) && !node.caseBlock.clauses.some(isDefaultClause)) {
+            if (!ctx.options.allowExhaustive) {
+                ctx.addFailureAtNode(node, Rule.FAILURE_STRING);
+            } else if (checker === undefined) {
+                throw Error("Expected `checker` to be defined");
+            } else {
+                const uncovered = uncoveredValues(checker, node);
+
+                if (uncovered === "non-exhaustive") {
+                    ctx.addFailureAtNode(node, Rule.FAILURE_STRING);
+                } else if (uncovered.length > 0) {
+                    ctx.addFailureAtNode(node, Rule.failureStringForMissingCases(uncovered));
+                }
+            }
         }
         return ts.forEachChild(node, cb);
     });
 }
+
+// returns a (possibly empty) array of missing values, or the string 'non-exhaustive' if the
```

Comment:
Nit: should be JSDoc-formatted `/* @returns` so that IDEs can surface this in hover information for the method. `//` comments won't be (at least in VS Code).